File: ast/statement/declaration.go
package statement

import (
	"fmt"

	"github.com/panda-io/micro-panda/ast"
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/expression"
)

type DeclarationStatement struct {
	StatementBase
	Name  *expression.Identifier
	Type  core.Type
	Value ast.Expression
}

func (d *DeclarationStatement) Validate(c ast.Context) {
	d.Type = c.ResolveType(d.Type)
	if d.Value != nil {
		d.Value.Validate(c, d.Type)
		if d.Value.Type() != nil && d.Type != nil && !d.Value.Type().Equal(d.Type) {
			c.Error(d.Value.GetPosition(), fmt.Sprintf("init value type mismatch with define, expect '%s', got '%s'", d.Type.String(), d.Value.Type().String()))
		}
	}
	if d.Type != nil {
		err := c.AddObject(d.Name.Name, d.Type)
		if err != nil {
			c.Error(d.GetPosition(), err.Error())
		}
	}
}

File: MicroPanda/AST/Statement/Declaration.cs
namespace MicroPanda.AST.Statement;

using Expression;
using Type;

internal class DeclarationStatement : Statement
{
    internal Identifier? Name { get; set; }
    internal Type? Type { get; set; }
    internal Expression? Value { get; set; }

    internal override void Validate(Context context)
    {
        Type = context.Program.ValidateType(Type!);
        if (Value != null)
        {
            Value.Validate(context, Type);
            if (Value.Type != null && Type != null && !Value.Type.Equal(Type))
            {
                context.Program.Error(Value.Position, "init value type mismatch with define");
            }
        }
        if (Type != null)
        {
            if(!context.AddObject(Name!.Name!, Type))
            {
                context.Program.Error(Name!.Position, $"redeclared variable: {Name.Name}");
            }
        }
    }
}